File: java/src/main/java/com/kale/interview/graphql/SubmissionDataFetcher.java
package com.kale.interview.graphql;

import com.kale.interview.data.Submission;
import com.kale.interview.services.SubmissionService;
import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsMutation;
import com.netflix.graphql.dgs.DgsQuery;
import com.netflix.graphql.dgs.InputArgument;

import java.util.List;

@DgsComponent
public class SubmissionDataFetcher {

    private final SubmissionService submissionService;

    public SubmissionDataFetcher(SubmissionService submissionService) {
        this.submissionService = submissionService;
    }

    @DgsQuery
    public List<Submission> submissions(@InputArgument String campaignId) {
        return submissionService.getSubmissionsByCampaign(Long.parseLong(campaignId));
    }

    @DgsMutation
    public Submission submitContent(@InputArgument SubmitContentInput input) {
        return submissionService.submitContent(
                Long.parseLong(input.campaignId()),
                input.creatorId(),
                input.contentUrl()
        );
    }

    @DgsMutation
    public Submission approveSubmission(@InputArgument String submissionId) {
        return submissionService.approveSubmission(Long.parseLong(submissionId));
    }

    @DgsMutation
    public Submission rejectSubmission(@InputArgument String submissionId) {
        return submissionService.rejectSubmission(Long.parseLong(submissionId));
    }

    public record SubmitContentInput(
            String campaignId,
            String creatorId,
            String contentUrl
    ) {}
}

File: kotlin/src/main/kotlin/com/kale/interview/graphql/SubmissionDataFetcher.kt
package com.kale.interview.graphql

import com.kale.interview.data.Submission
import com.kale.interview.services.SubmissionService
import com.netflix.graphql.dgs.DgsComponent
import com.netflix.graphql.dgs.DgsMutation
import com.netflix.graphql.dgs.DgsQuery
import com.netflix.graphql.dgs.InputArgument

@DgsComponent
class SubmissionDataFetcher(private val submissionService: SubmissionService) {

    @DgsQuery
    fun submissions(@InputArgument campaignId: String): List<Submission> =
        submissionService.getSubmissionsByCampaign(campaignId.toLong())

    @DgsMutation
    fun submitContent(@InputArgument input: SubmitContentInput): Submission =
        submissionService.submitContent(
            campaignId = input.campaignId.toLong(),
            creatorId = input.creatorId,
            contentUrl = input.contentUrl,
        )

    @DgsMutation
    fun approveSubmission(@InputArgument submissionId: String): Submission =
        submissionService.approveSubmission(submissionId.toLong())

    @DgsMutation
    fun rejectSubmission(@InputArgument submissionId: String): Submission =
        submissionService.rejectSubmission(submissionId.toLong())

    @DgsMutation
    fun processPayment(@InputArgument submissionId: String): Submission =
        submissionService.processPayment(submissionId.toLong())
}

data class SubmitContentInput(
    val campaignId: String,
    val creatorId: String,
    val contentUrl: String,
)
